metrics-v2/src/dynmetrics.rs

```diff
@@ -0,0 +1,187 @@
+/// // Normally you would use some utility to do this. (e.g. pin-utils)
+/// let my_dyn_metric = unsafe { Pin::new_unchecked(&my_dyn_metric) };
+/// my_dyn_metric.register("a.dynamic.counter");
+///
+/// let metrics = metrics();
+/// assert_eq!(metrics.dynamic_metrics()[0].name(), "a.dynamic.counter");
+/// ```
+///
+/// [`register`]: crate::dynmetrics::DynPinnedMetric::register
+pub struct DynPinnedMetric<M: Metric> {
+    metric: M,
+    // This type relies on Pin's guarantees for correctness. Allowing it to be unpinned would cause
+    // errors.
+    _marker: PhantomPinned,
+}
+
+impl<M: Metric> DynPinnedMetric<M> {
+    /// Create a new `DynPinnedMetric` with the provided internal metric.
+    ///
+    /// This does not register the metric. To do that call [`register`].
+    ///
+    /// [`register`]: self::DynPinnedMetric::register
+    pub fn new(metric: M) -> Self {
+        Self {
+            metric,
+            _marker: PhantomPinned,
+        }
+    }
+
+    /// Register this metric in the global list of dynamic metrics with `name`.
+    ///
+    /// Calling this multiple times will result in the same metric being
+    /// registered multiple times under potentially different names.
+    pub fn register(self: Pin<&Self>, name: impl Into<Cow<'static, str>>) {
+        // SAFETY:
+        // To prove that this is safe we need to list out a few guarantees/requirements:
+        //  - Pin ensures that the memory of this struct instance will not be reused
+        //    until the drop call completes.
+        //  - MetricEntry::new_unchecked requires that the metric reference outlive
+        //    created the MetricEntry instance.
+        //
+        // Finally, register, will keep the MetricEntry instance in a global list until
+        // the corresponding unregister call is made.
+        //
+        // Taking all of these together, we can guarantee that self.metric will not be
+        // dropped until this instance of DynPinnedMetric is dropped itself. At that
+        // point, drop calls unregister which will drop the MetricEntry instance. This
+        // ensures that the references to self.metric in REGISTRY will always be valid
+        // and that this method is safe.
+        unsafe { register(MetricEntry::new_unchecked(&self.metric, name.into())) };
+    }
+}
+
+impl<M: Metric> Drop for DynPinnedMetric<M> {
+    fn drop(&mut self) {
+        // If this metric has not been registered then nothing will be removed.
+        unregister(&self.metric);
+    }
+}
+
+impl<M: Metric> Deref for DynPinnedMetric<M> {
+    type Target = M;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        &self.metric
+    }
+}
+
+/// A dynamic metric that stores the metric instance on the heap.
+///
+/// This avoids a lot of the hangup with [`DynPinnedMetric`] as it allows for
+/// moving the `DynBoxedMetric` without having to worry about pinning or safety
+/// issues. However, this comes at the expense of requiring a heap allocation
```

Comment:
Are we expecting any performance difference? If so, some benchmarks would be nice.

Reply:
Once the metric is created there shouldn't be any performance difference (barring cache effects which will be rather hard to make show up in a benchmark). However, creating a `DynBoxedMetric` requires creating an allocation on the heap which will be more expensive than just initializing an atomic on the stack.